### PR TITLE
Add cache strategy that uses the new caching API

### LINKED ISSUE #3345: Skip analysis for unchanged files when the analysis allows it
When the sensor context indicates that unchanged files can be skipped from analysis (in practice this is during PR analysis), the SonarJS analyzers should execute only the UCFG rule (from sonar-security frontend) for files marked as unchanged.

UCFG rule generates UCFG files required for security tainted analysis. The eslint-bridge loads this external rule dynamically according to the 'bundles' node command parameter and then the ESLint linter is created based on the request rules parameter.
Thus a second 'optimized' linter should be created in the initialization service with only this external rule (`ucfg`). And the JS/TS/YAML analysis services need a new flag triggering the use of the 'optimized' linter.

On the integration test side we need to validate that UCFG are properly generated for unchanged files. For that we need to add sonar-security plugin to the orchestrator and assert presence of either UCFG files (low level check) or security issues (then some security rule should be activated). Probably we should go for a first option as more reliable and controllable by us.

### LINKED ISSUE #3362: Use the plugin caching API to store and retrieve UCFG files
In order to speed up the analysis, the plugin API provides a caching API to reuse content generated by sensors which can be applied to UCFG files.
On the java plugin side, the sensors should leverage the caching API to populate the cache and skip completely the call to the eslint-bridge for unchanged files for which there are cached UCFG files.
On the javascript side, analyzers should return the paths to the UCFG files generated during the analysis.

### PR DESCRIPTION
**Pull request by francoismora**
Fixes #3362

### CODE REVIEW

**Review by saberduck — CHANGES_REQUESTED** [review]

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 1]
```diff
@@ -128,8 +147,14 @@ void should_analyse_cloudformation_pull_requests() {
         .logsAtLeastOnce("DEBUG: Files which didn't change will be part of UCFG generation only, other rules will not be executed")
         .logsOnce("DEBUG: Initializing linter \"default\"")
         .logsOnce("DEBUG: Initializing linter \"unchanged\"")
-        .logsOnce("file1.yaml\" with linterId \"unchanged\"")
+        .logsOnce("Cache strategy set to 'READ_AND_WRITE' for file 'file1.yaml'")
```
[the remark this answers is not in the document]
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:150]: This change will be implemented in a separate PR

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintBridgeServer.java [hunk 2]
```diff
@@ -104,6 +104,7 @@ class AnalysisResponse {
     Metrics metrics = new Metrics();
     CpdToken[] cpdTokens = {};
     Perf perf;
+    String[] ucfgPaths = {};
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintBridgeServer.java:107]: This change will be implemented in a separate PR

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheSerialization.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+import org.sonar.api.batch.sensor.cache.WriteCache;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+
+class CacheSerialization {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheSerialization.java:54]: This change will be implemented in a separate PR

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      var jsonCacheKey = createCacheKey(context, "JSON", inputFile);
+      var manifest = JSON.readCache(context.previousCache(), jsonCacheKey, null);
+      context.nextCache().copyFromPrevious(jsonCacheKey);
+
+      var sequenceCacheKey = createCacheKey(context, "SEQ", inputFile);
+      var config = new CacheSerialization.SequenceConfig(getWorkingDirectoryAbsolutePath(context), manifest);
+      SEQUENCE.readCache(context.previousCache(), sequenceCacheKey, config);
+      context.nextCache().copyFromPrevious(sequenceCacheKey);
+
+      success = true;
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java:103]: This change will be implemented in a separate PR

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      log(CacheStrategy.READ_AND_WRITE, inputFile, null);
+      return CacheStrategy.READ_AND_WRITE;
+    }
+
+    if (!canSkipUnchangedFiles) {
+      log(CacheStrategy.WRITE_ONLY, inputFile, "current analysis requires all files to be analyzed");
+    } else if (!isFileUnchanged) {
+      log(CacheStrategy.WRITE_ONLY, inputFile, "the current file is changed");
+    } else {
+      log(CacheStrategy.WRITE_ONLY, inputFile, "the current file is not cached");
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java:139]: Actually the message corresponds to the case where there's no cache entry for the file (`!isFileInCache`).

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+  boolean isAnalysisRequired(SensorContext context, InputFile inputFile) {
+    if (this != CacheStrategy.READ_AND_WRITE) {
+      return true;
+    }
+    boolean isCacheReadSuccessful = readAndWrite(context, inputFile);
+    return !isCacheReadSuccessful;
+  }
+
+  void writeGeneratedFilesToCache(SensorContext context, InputFile inputFile, @Nullable String[] generatedFiles) throws IOException {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java:155]: This change will be implemented in a separate PR

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    } else if (!isFileUnchanged) {
+      log(CacheStrategy.WRITE_ONLY, inputFile, "the current file is changed");
+    } else {
+      log(CacheStrategy.WRITE_ONLY, inputFile, "the current file is not cached");
+    }
+
+    return CacheStrategy.WRITE_ONLY;
+  }
+
+  private static final Logger LOG = Loggers.get(AnalysisMode.class);
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java]: Changed

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.utils.Version;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import static org.sonar.plugins.javascript.eslint.CacheSerialization.json;
+import static org.sonar.plugins.javascript.eslint.CacheSerialization.sequence;
+
+enum CacheStrategy {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java:35]: This change will be implemented in a separate PR

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  WRITE_ONLY, READ_AND_WRITE, NO_CACHE;
+
+  static final CacheSerialization.JsonSerialization<CacheSerialization.FilesManifest> JSON = json(CacheSerialization.FilesManifest.class);
+  static final CacheSerialization.SequenceSerialization SEQUENCE = sequence();
+
+  private static Path getWorkingDirectoryAbsolutePath(SensorContext context) {
+    return context.fileSystem().workDir().toPath();
+  }
+
+  private static String createCacheKey(SensorContext context, String category, InputFile inputFile) {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/CacheStrategy.java]: This change will be implemented in a separate PR

**Review by saberduck — APPROVED** [review]

**Pull request merged**